File: src/main/java/io/javalin/serversentevent/EventSourceImpl.java
package io.javalin.serversentevent;

public class EventSourceImpl implements EventSource {

    private Emitter emitter;
    private SSEClose close;

    public EventSourceImpl(Emitter emitter) {
        this.emitter = emitter;
    }

    @Override
    public void onOpen(SSEConnect open) {
        open.handler( this );
    }

    @Override
    public void onClose(SSEClose close) {
        this.close = close;
    }

    @Override
    public void sendEvent(String event, String data) {
        emitter.event( event, data );
        if(emitter.isClose()) {
            close.handler( this );
        }
    }
}

File: src/main/java/io/javalin/serversentevent/EventSourceImpl.kt
package io.javalin.serversentevent

class EventSourceImpl(private val emitter: Emitter, private val pathParam: Map<String, String>) : EventSource {

    private var close: SSEClose? = null

    override fun onOpen(open: SSEConnect) {
        open(this)
    }

    override fun onClose(close: SSEClose) {
        this.close = close
    }

    override fun sendEvent(event: String, data: String) {
        emitter.event(event, data)
        if (emitter.isClose()) {
            close?.invoke(this)
        }
    }

    override fun pathParamMap(): Map<String, String> {
        return pathParam
    }
}
